### PR TITLE
Check base tablet when clean alter task during startup

### PR DESCRIPTION
**Pull request by yiguolei**

### CODE REVIEW

File: be/src/olap/tablet_manager.cpp [hunk 1]
```diff
@@ -239,6 +239,15 @@ void TabletManager::cancel_unfinished_schema_change() {
             }
 
             AlterTabletTaskSharedPtr new_alter_task = new_tablet->alter_task();
+            if (new_alter_task != nullptr 
+                && (new_alter_task->related_tablet_id() != tablet->tablet_id() 
+                    || new_alter_task->related_schema_hash() != tablet->schema_hash())) {
+                LOG(WARNING) << "base tablet " << tablet->full_name()
+                             << "new tablet " << new_tablet->full_name()
```
Comment (kangpinghuang) [inline be/src/olap/tablet_manager.cpp]:
```suggestion
                             << " new tablet " << new_tablet->full_name()
```
add a space. The same the the following line.

File: be/src/olap/tablet_manager.cpp [hunk 2]
```diff
@@ -258,18 +267,28 @@ void TabletManager::cancel_unfinished_schema_change() {
...
-            }
-            res = new_tablet->save_meta();
-            if (res != OLAP_SUCCESS) {
-                LOG(FATAL) << "fail to save new tablet meta. res=" << res
-                           << ", new_tablet=" << new_tablet->full_name();
-                return;
+            if (new_alter_task == nullptr 
+                && new_tablet->creation_time() < tablet->creation_time()) {
+                // case 1: create new tablet and save meta successfully, but failed to save alter state in base tablet
+                // case 2: during clear stage, clear base successfully, but faile dto clear new tablet
```
Comment (kangpinghuang) [inline be/src/olap/tablet_manager.cpp]:
```suggestion
                // case 2: during clear stage, clear base successfully, but failed to clear new tablet
```

File: be/src/olap/tablet_manager.cpp [hunk 3]
```diff
@@ -258,18 +267,28 @@ void TabletManager::cancel_unfinished_schema_change() {
...
+                LOG(WARNING) << "base tablet's alter task is null, skip set state"
+                             << " base_tablet=" << new_tablet->full_name()
+                             << " create_time=" << new_tablet->creation_time()
+                             << " new_tablet=" << tablet->full_name()
+                             << " create_time=" << tablet->creation_time();
+            } else {
+                res = new_tablet->set_alter_state(ALTER_FAILED);
+                if (res != OLAP_SUCCESS) {
+                    LOG(FATAL) << "fail to save new tablet meta. res=" << res
```
Comment (kangpinghuang) [inline be/src/olap/tablet_manager.cpp]: modify the failure message. I think failed to set the alter state is better.

**Review by kangpinghuang — CHANGES_REQUESTED** [review]

File: be/src/olap/tablet_manager.cpp [hunk 4]
```diff
@@ -258,18 +267,28 @@ void TabletManager::cancel_unfinished_schema_change() {
...
+                && new_tablet->creation_time() < tablet->creation_time()) {
+                // case 1: create new tablet and save meta successfully, but failed to save alter state in base tablet
+                // case 2: during clear stage, clear base successfully, but faile dto clear new tablet
+                LOG(WARNING) << "base tablet's alter task is null, skip set state"
+                             << " base_tablet=" << new_tablet->full_name()
+                             << " create_time=" << new_tablet->creation_time()
+                             << " new_tablet=" << tablet->full_name()
+                             << " create_time=" << tablet->creation_time();
+            } else {
+                res = new_tablet->set_alter_state(ALTER_FAILED);
```
Comment (kangpinghuang) [inline be/src/olap/tablet_manager.cpp]: what if new_alter_task == nullptr?

**Review by kangpinghuang — CHANGES_REQUESTED** [review]

File: be/src/olap/tablet_manager.cpp (hunk 4 above)
Comment (yiguolei) [reply be/src/olap/tablet_manager.cpp]: fatal error

File: be/src/olap/tablet_manager.cpp [hunk 5]
```diff
@@ -239,6 +239,15 @@ void TabletManager::cancel_unfinished_schema_change() {
             }
 
             AlterTabletTaskSharedPtr new_alter_task = new_tablet->alter_task();
+            if (new_alter_task != nullptr 
+                && (new_alter_task->related_tablet_id() != tablet->tablet_id() 
+                    || new_alter_task->related_schema_hash() != tablet->schema_hash())) {
```
Comment (kangpinghuang) [inline be/src/olap/tablet_manager.cpp]: indent with the last line

**Review by kangpinghuang — APPROVED** [review]

**Pull request merged**